### PR TITLE
N-API Support

### PR DESCRIPTION
**Pull request by jschlight**
This pull request implements changes to `prebuild` to support building and uploading Node native modules based on -API by adding a new `napi` runtime. The changes are summarized on the [README](https://github.com/jschlight/prebuild/tree/prebuild-napi#n-api-considerations).

This PR should be coordinated with its [counterpart PR](https://github.com/prebuild/prebuild-install/pull/87) on `prebuild-install`.

### CODE REVIEW

File: napi.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
+'use strict'
```
[the remark this answers is not in the document]
Comment (jschlight) [reply napi.js]: Yes. I hadn't thought of that.

**Comment by jschlight** [comment]
I've completed all planned work on this PR. Are there further concerns that need to be addressed before it can be merged?

**Comment by ralphtheninja** [comment]
Tests are failing locally, restarting tests on travis to see what's going on.

**Comment by ralphtheninja** [comment]
Not failing on Travis. Maybe something weird on my system.

```
# can prebuild a cmake-js native module for node

not ok 166 should be equal
  ---
    operator: equal
    expected: |-
      null
    actual: |-
      { [Error: Command failed: npm run prebuild
prebuild info begin Prebuild version 8.0.1
prebuild info build Preparing to prebuild native@1.0.0 for node 8.11.1 on linux-x64 using cmake-js
info TOOL Using Ninja generator, because ninja is available.
info TOOL Using c++11 compiler standard.
ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
prebuild ERR! build Error: Failed to build cmake with exit code 1
prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! native@1.0.0 prebuild: `node ../../bin.js --backend cmake-js --target 8.11.1 --runtime node`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the native@1.0.0 prebuild script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_39_543Z-debug.log
] killed: false, code: 2, signal: null, cmd: 'npm run prebuild' }
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:16:7)
    stack: |-
      Error: Command failed: npm run prebuild
      prebuild info begin Prebuild version 8.0.1
      prebuild info build Preparing to prebuild native@1.0.0 for node 8.11.1 on linux-x64 using cmake-js
      info TOOL Using Ninja generator, because ninja is available.
      info TOOL Using c++11 compiler standard.
      ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
      prebuild ERR! build Error: Failed to build cmake with exit code 1
      prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
      prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
      prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
      npm ERR! code ELIFECYCLE
      npm ERR! errno 2
      npm ERR! native@1.0.0 prebuild: `node ../../bin.js --backend cmake-js --target 8.11.1 --runtime node`
      npm ERR! Exit status 2
      npm ERR! 
      npm ERR! Failed at the native@1.0.0 prebuild script.
      npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
      npm WARN Local package.json exists, but node_modules missing, did you mean to install?
      
      npm ERR! A complete log of this run can be found in:
      npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_39_543Z-debug.log
      
          at ChildProcess.exithandler (child_process.js:289:12)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...
not ok 167 should be equal
  ---
    operator: equal
    expected: true
    actual:   false
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:17:7)
    stack: |-
      Error: should be equal
          at Test.assert [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:224:54)
          at Test.bound [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at Test.equal (/home/lms/src/prebuild/node_modules/tape/lib/test.js:384:10)
          at Test.bound [as equal] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at /home/lms/src/prebuild/test/integration-cmake-test.js:17:7
          at ChildProcess.exithandler (child_process.js:296:5)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...
# can prebuild a cmake-js native module for electron

not ok 168 should be equal
  ---
    operator: equal
    expected: |-
      null
    actual: |-
      { [Error: Command failed: npm run prebuild-electron
prebuild info begin Prebuild version 8.0.1
prebuild info build Preparing to prebuild native@1.0.0 for electron 1.4.11 on linux-x64 using cmake-js
info TOOL Using Ninja generator, because ninja is available.
info TOOL Using c++11 compiler standard.
ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
prebuild ERR! build Error: Failed to build cmake with exit code 1
prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! native@1.0.0 prebuild-electron: `node ../../bin.js --backend cmake-js --target 1.4.11 --runtime electron`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the native@1.0.0 prebuild-electron script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_40_521Z-debug.log
] killed: false, code: 2, signal: null, cmd: 'npm run prebuild-electron' }
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:29:7)
    stack: |-
      Error: Command failed: npm run prebuild-electron
      prebuild info begin Prebuild version 8.0.1
      prebuild info build Preparing to prebuild native@1.0.0 for electron 1.4.11 on linux-x64 using cmake-js
      info TOOL Using Ninja generator, because ninja is available.
      info TOOL Using c++11 compiler standard.
      ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
      prebuild ERR! build Error: Failed to build cmake with exit code 1
      prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
      prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
      prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
      npm ERR! code ELIFECYCLE
      npm ERR! errno 2
      npm ERR! native@1.0.0 prebuild-electron: `node ../../bin.js --backend cmake-js --target 1.4.11 --runtime electron`
      npm ERR! Exit status 2
      npm ERR! 
      npm ERR! Failed at the native@1.0.0 prebuild-electron script.
      npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
      npm WARN Local package.json exists, but node_modules missing, did you mean to install?
      
      npm ERR! A complete log of this run can be found in:
      npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_40_521Z-debug.log
      
          at ChildProcess.exithandler (child_process.js:289:12)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...
not ok 169 should be equal
  ---
    operator: equal
    expected: true
    actual:   false
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:30:7)
    stack: |-
      Error: should be equal
          at Test.assert [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:224:54)
          at Test.bound [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at Test.equal (/home/lms/src/prebuild/node_modules/tape/lib/test.js:384:10)
          at Test.bound [as equal] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at /home/lms/src/prebuild/test/integration-cmake-test.js:30:7
          at ChildProcess.exithandler (child_process.js:296:5)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...
# can prebuild a cmake-js native module for node with silent argument

ok 170 stdout should be silent
not ok 171 should be equal
  ---
    operator: equal
    expected: |-
      null
    actual: |-
      { [Error: Command failed: npm run prebuild-silent
prebuild info begin Prebuild version 8.0.1
prebuild info build Preparing to prebuild native@1.0.0 for node 8.11.1 on linux-x64 using cmake-js
info TOOL Using Ninja generator, because ninja is available.
info TOOL Using c++11 compiler standard.
ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
prebuild ERR! build Error: Failed to build cmake with exit code 1
prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! native@1.0.0 prebuild-silent: `node ../../bin.js --backend cmake-js --target 8.11.1 --runtime node -- -i`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the native@1.0.0 prebuild-silent script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_41_438Z-debug.log
] killed: false, code: 2, signal: null, cmd: 'npm run prebuild-silent' }
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:43:7)
    stack: |-
      Error: Command failed: npm run prebuild-silent
      prebuild info begin Prebuild version 8.0.1
      prebuild info build Preparing to prebuild native@1.0.0 for node 8.11.1 on linux-x64 using cmake-js
      info TOOL Using Ninja generator, because ninja is available.
      info TOOL Using c++11 compiler standard.
      ERR! OMG CMake executable is not found. Please use your system's package manager to install it, or you can get installers from there: http://cmake.org.
      prebuild ERR! build Error: Failed to build cmake with exit code 1
      prebuild ERR! build     at ChildProcess.<anonymous> (/home/lms/src/prebuild/cmakebuild.js:29:19)
      prebuild ERR! build     at ChildProcess.emit (events.js:182:13)
      prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:239:12)
      npm ERR! code ELIFECYCLE
      npm ERR! errno 2
      npm ERR! native@1.0.0 prebuild-silent: `node ../../bin.js --backend cmake-js --target 8.11.1 --runtime node -- -i`
      npm ERR! Exit status 2
      npm ERR! 
      npm ERR! Failed at the native@1.0.0 prebuild-silent script.
      npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
      npm WARN Local package.json exists, but node_modules missing, did you mean to install?
      
      npm ERR! A complete log of this run can be found in:
      npm ERR!     /home/lms/.npm/_logs/2018-09-27T19_06_41_438Z-debug.log
      
          at ChildProcess.exithandler (child_process.js:289:12)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...
not ok 172 should be equal
  ---
    operator: equal
    expected: true
    actual:   false
    at: <anonymous> (/home/lms/src/prebuild/test/integration-cmake-test.js:44:7)
    stack: |-
      Error: should be equal
          at Test.assert [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:224:54)
          at Test.bound [as _assert] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at Test.equal (/home/lms/src/prebuild/node_modules/tape/lib/test.js:384:10)
          at Test.bound [as equal] (/home/lms/src/prebuild/node_modules/tape/lib/test.js:76:32)
          at /home/lms/src/prebuild/test/integration-cmake-test.js:44:7
          at ChildProcess.exithandler (child_process.js:296:5)
          at ChildProcess.emit (events.js:182:13)
          at maybeClose (internal/child_process.js:961:16)
          at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
  ...

1..172
# tests 172
# pass  166
# fail  6

```

**Comment by ralphtheninja** [comment]
Nvm, I was missing `cmake` on my machine, which is installed on Travis by default. So this was indeed an issue on my machine only.

**Comment by ralphtheninja** [comment]
@jschlight Thanks for your contribution and sorry for delaying this! Released in 8.1.0.

**Comment by ralphtheninja** [comment]
@jschlight We could use you in prebuild org, if you are interested in an invite :smile:

**Comment by jschlight** [comment]
@ralphtheninja Yes. Please send the invite.

**Pull request merged**